pkg/controller/container-runtime-config/helpers.go

```diff
@@ -264,6 +297,49 @@ func updateRegistriesConfig(data []byte, internalInsecure, internalBlocked []str
 	return newData.Bytes(), nil
 }
 
+// updatePolicyJSON decodes the data rendered from the template, merges the changes in and encodes it
+// back into a JSON format. It returns the bytes of the encoded data
+// It also returns an error if both allowed and blocked registries are set
+func updatePolicyJSON(data []byte, internalBlocked, internalAllowed []string) ([]byte, error) {
+	if len(internalAllowed) != 0 && len(internalBlocked) != 0 {
+		return nil, fmt.Errorf("invalid images config: only one of AllowedRegistries or BlockedRegistries may be specified")
+	}
+
+	policyObj := &signature.Policy{}
```

Comment:
This is more of a question than a review comment:

`updateRegistriesConfig` decodes the original config, and only updates it (well… partially; it just overwrites some values regardless of whether the previous values were MCO-written or got there some other way). Does this need to do the same thing? Can `updateRegistriesConfig` be simplified to generate a fresh configuration instead of reading the old one? Or, maybe, are the two files somehow different to warrant different treatment?

(I rather hope that overwriting from scratch is OK; it’s not quite obvious how to update a pre-existing policy to implement `internalBlocked`/`internalAllowed`.)

Reply:
… reading more of the code, if I understand it correctly,
> `updateRegistriesConfig` decodes the original config
It does not decode the “current on-disk status” (with possible custom edits); it decodes the result of processing the built-in `templates/`, which would otherwise be used. `updateRegistriesConfig` then creates a new `MachineConfig` object, and the render controller ensures that the last version (by name) is applied, effectively overwriting the template output with the result of `updateRegistriesConfig` (template).

This means that `updateRegistriesConfig`, strictly speaking, needs to only handle updates to the template file, not to arbitrary files.

OK, that’s easier.

---

Per the other review comments, the `policy.json` generation, in that case, does not quite follow the existing pattern AFAICS.

Reply:
Either way, the generation of `getManagedKeyReg`-named `MachineConfig` which includes `policy.json` will make it **rather difficult** for users to use custom `policy.json` configurations (e.g. to actually specify trusted keys and signature enforcement).  I guess there is always `999-*` and the like for custom configurations, at the cost of sacrificing the whitelist/blacklist functionality.  Is that good enough?

Reply:
Yeah, `updateRegistriesConfig` just updates the registries.conf files based on the values sent in the CRD. That is what the MCO does for crio.conf and storage.conf as well.
So `updateRegistriesConfig` is just for reading in the registries.conf file and updating it. `updatePolicyJSON` is just for policy.json.
Do we want to create a new policy.json from scratch every time a change happens?

Reply:
> Do we want to create a new policy.json from scratch every time a change happens?
 
(I am new to this code.) AFAICS nodes always use the configuration in `MachineConfigPoolStatus.Configuration`, which is only generated by the render controller, so the files in `templates/*` are never used directly/unmodified; `container-runtime-config.Controller.syncImageConfig` is always involved. Therefore, it does not really matter much whether there is a template that is updated by `updatePolicyJSON`, or whether there is no template and this function generates a configuration from scratch.

Still, the code should behave consistently whether or not `blockedRegs != nil || allowedRegs != nil`; currently if this is false, the template is used with the `docker-daemon` entry, and if this is false, the template is ignored and there is no `docker-daemon` entry. That would be a clearly unintended side effect of the `blockedRegs`/`allowedRegs` configuration.

---

It seems simplest to me to drop the template and to always generate the policy using this function; that way all the decisions are clearly located in one place.

Using the template-created `data` object in this function would also work (whether or not the template removes the `docker-daemon` entry… but seems a bit more risky: the `policyObj.Default` `PolicyRequirements` value has fundamental impact to whether other, more specific, scopes, are configured correctly or not (e.g. whether there needs to exist a `reject` entry for a scope or whether the default would catch it) - so giving future maintainers a `policy.json` template that they can edit without necessarily noticing that the default policy can change is a bit more risky. (OTOH, editing the template JSON is a bit more readable than editing the Go code in this function; on balance, I don’t see this as a strong enough argument, but opinions can differ.)

Reply:
@runcom WDYT? Is it okay if we just generate the policy.json without having a template from before every time there is an allowed or blocked registry entry in the images CR?